Reference implementations for rsqrt and native_layer_norm

IvanYashchuk opened the pull request:
This PR adds references for:
- `torch.rsqrt`
- `torch.native_layer_norm`
-  `torch.nn.functional.layer_norm`

`native_layer_norm` had a different number of dimensions if the input was 0-sized. I fixed that.

IvanYashchuk committed:
Remove special casing for 0-sized inputs

IvanYashchuk committed:
Add OpInfo for native_layer_norm

IvanYashchuk committed:
Remove special casing for 0-sized in decompositions.py

IvanYashchuk committed:
Skip CPU bfloat16 for native_layer_norm

IvanYashchuk committed:
Add refs for rsqrt, native_layer_norm

IvanYashchuk committed:
Add samples with None arguments for weight and bias

IvanYashchuk committed:
Three cases for optional weight and bias

IvanYashchuk committed:
Add error_inputs_native_layer_norm

IvanYashchuk committed:
Add input checks

IvanYashchuk committed:
Remove native_layer_norm decomp

IvanYashchuk committed:
formatting

IvanYashchuk committed:
Skip mypy

vadimkantorov commented:
It would be nice to have a way to display reference implementation code in documentation, this can compensate for unclear description (meanwhile it's being improved): https://github.com/pytorch/pytorch/issues/51455

ngimel commented:
cc @chillee for native_layer_norm

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -1762,6 +1768,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
+def _normalize(a, norm_dims, eps):
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a_acc = prims.convert_element_type(a, computation_dtype)
+    biased_var, mean = var_mean(a_acc, dim=norm_dims, unbiased=False, keepdim=True)
```
currently nvfuser doesn't understand var_mean, so this decomp is not readily usable

Chillee replied:
Same here.

Also, the reason the previous decomp used separate calls to `var` and `mean` was that it more closely matched the numerics of eager mode. @ngimel's opinion was that the discrepancy shouldn't matter, but just noting in case you need to adjust tolerances on any tests.

IvanYashchuk replied:
I will make `mean` working with the nvFuser executor in a separate PR. `var` will be implemented later as it's currently a "prim" in PyTorch, but a composite in nvFuser.

It's separate calls to `var` and `mean`. Take a look at the `var_mean` function:
```
def var_mean(
    a: TensorLikeType,
    dim: Union[Optional[int], Optional[List[int]]] = None,
    unbiased: Optional[bool] = None,
    keepdim: bool = False,
    *,
    correction: Optional[int] = None,
):
    v = var(a, dim, unbiased, keepdim, correction=correction)
    m = mean(a, dim, keepdim)
    return v, m
```

IvanYashchuk replied:
PR for mean: https://github.com/pytorch/pytorch/pull/79444
PR for var: https://github.com/pytorch/pytorch/pull/79517
And together they make nvfuser understand var_mean.

Lezcano replied:
I believe @peterbell10 once told me that `var_mean` had some speed and stability problems in PyTorch (?). I do not remember very well so I'll let her discuss this, but perhaps this is relevant here.

peterbell10 replied:
On CPU, the mean from `var_mean` is less accurate than calling `mean` separately. `mean` is implemented roughly as `sum(x, dim) / x.size(dim)`, where `sum` uses a low-error summation algorithm. `var_mean` on the other hand computes both the mean and var in a single pass, but with a naive summation that is less accurate.

Chillee replied:
For similar reasons, can we change this call to a `torch.var_mean` as well?

ngimel replied:
Well, then, unless you decompose `torch.var_mean` further, you'd get inaccurate cpu result

Chillee commented on test/test_decomp.py:
```diff
@@ -450,6 +450,9 @@ def check_decomposed(aten_name):
         func = op.get_op()
         for sample_input in samples:
             if requires_grad:
+                if None in sample_input.args:
```
Why is this needed?

Chillee commented on torch/_refs/__init__.py:
```diff
@@ -1762,6 +1768,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a_acc = prims.convert_element_type(a, computation_dtype)
+    biased_var, mean = var_mean(a_acc, dim=norm_dims, unbiased=False, keepdim=True)
+    rstd = rsqrt(biased_var + eps)
```
Can we just call `torch.rsqrt` here?

Chillee replied:
My understanding is that the current thinking is to prefer calling `torch.foo` vs. `_refs.foo` if convenient, as it's fairly easy to map from `torch.foo` => `_refs.foo` but not the other way around.

See @mruberry's comment here: https://github.com/pytorch/pytorch/pull/78689#discussion_r889159026

IvanYashchuk replied:
It's very easy to map `_refs.rsqrt` to `torch.rsqrt` because `_refs.rsqrt` is a wrapper of `torch.rsqrt` for regular and meta tensors and `test_decompy.py` passing indicates that it's fine as it is.

I changed this to `torch.rsqrt`.

Chillee replied:
From the perspective of the "correctness" tests it's fine. But, if I understand correctly, `_refs.rsqrt => prims.rsqrt`, and then you're saying that `prims.rsqrt` simply calls `torch.rsqrt` under the hood. 

But when we trace things out, using `_refs.rsqrt` doesn't have an interception point to allow us to decompose `native_layer_norm` but not decompose `_refs.rsqrt`.

TBH, for `rsqrt` this is a somewhat strange discussion haha since `_refs.rsqrt == prims.rsqrt == aten::rsqrt`. But it matters more for `var_mean`.

For example, say that I'm a compiler that wants `native_layer_norm` decomposed, but I have a special handling for `var_mean`. If we call `_refs.var_mean`, then our trace has no option to trace out `var_mean` - we'll get `var` followed by `mean`.

If we use `torch.var_mean` though, then we'll have the choice to either further decompose into `var` and `mean` or preserve `var_mean` in our trace.

mruberry replied:
Because of the way the current context mapping is implemented it's a one-way torch. -> refs. transform, so we should get in the habit of preferring torch. calls when possible. In some cases the torch operation doesn't have as much functionality as the ref, and in cases where that extra functionality is used the ref should be called explicitly, instead.

Chillee commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -13445,6 +13519,23 @@ def error_inputs_mean(op_info, device, **kwargs):
+               # IndexError: tuple index out of range
+               DecorateInfo(unittest.skip('Skipped!'), 'TestGradients', 'test_forward_mode_AD'),
+               # Difference from float64 is larger with decomposition native_layer_norm_backward.default than original
+               DecorateInfo(unittest.skip('Skipped!'), 'TestDecomp', 'test_comprehensive',
```
yeah actually, this might be what you're seeing.

IvanYashchuk committed:
Remove test_comprehensive skip

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into native-layer-norm

IvanYashchuk committed:
Revert "Remove test_comprehensive skip"

This reverts commit 87cdca6752e636c6664e5b80552fb34cc7e2c4fe.

IvanYashchuk committed:
Remove test_comprehensive skip

IvanYashchuk committed:
Add rsqrt to __all__

IvanYashchuk committed:
Use torch.rsqrt per Horace's request

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into native-layer-norm

mruberry commented:
> It would be nice to have a way to display reference implementation code in documentation, this can compensate for unclear description (meanwhile it's being improved): #51455

https://github.com/pytorch/pytorch/issues/79518

mruberry commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -852,21 +852,21 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_cuda(
   if (M > 0) {
     LayerNormKernelImpl(*X, *gamma, *beta, M, N, eps, &Y, &mean, &rstd);
+  }
+  const auto input_shape = input.sizes();
```
Let's add a comment for this block -- is this a BC-breaking change?

IvanYashchuk replied:
> is this a BC-breaking change?
Yes, it does change the shape of var and mean returned for input tensors with 0s in the shape. It's a bug-fix.

IvanYashchuk replied:
Behavior on master:
```py
In [1]: import torch

In [2]: a = torch.randn(2, 0, 3, 3, device='cuda') # note 0 in the 2nd dim

In [3]: [x.shape for x in torch.native_layer_norm(a, (3, 3), None, None, 1e-5)]
Out[3]: [torch.Size([2, 0, 3, 3]), torch.Size([0]), torch.Size([0])] # shapes for var and mean are wrong

In [4]: a = torch.randn(2, 1, 3, 3, device='cuda') # note non-zero in the 2nd dim

In [5]: [x.shape for x in torch.native_layer_norm(a, (3, 3), None, None, 1e-5)]
Out[5]: [torch.Size([2, 1, 3, 3]), torch.Size([2, 1, 1, 1]), torch.Size([2, 1, 1, 1])]
```

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1808,6 +1815,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
 
+def _normalize(a, norm_dims, eps):
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a_acc = prims.convert_element_type(a, computation_dtype)
```
Let's make this conversation conditional

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1808,6 +1815,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
     return prims.slice_in_dim(a, start, start + length, axis=dim)
 
 
+def _normalize(a, norm_dims, eps):
```
Add a comment for this function -- what it's meant to be used for, how it's intended to be used -- we should probably add type annotations to it

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1808,6 +1815,75 @@ def narrow(a: TensorLikeType, dim: int, start: int, length: int) -> TensorLikeTy
+    weight: Optional[Tensor],
+    bias: Optional[Tensor],
+    eps: float,
+) -> Tuple[TensorLikeType, TensorLikeType, TensorLikeType]:
```
With the latest updates I think it's just all `Tensor` now

IvanYashchuk committed:
Add a comment to layer_norm_kernel.cu

IvanYashchuk committed:
Use _maybe_convert_to_dtype

IvanYashchuk committed:
Add docstring to _normalize

IvanYashchuk committed:
Change TensorLikeType->Tensor

IvanYashchuk committed:
Fix typo

IvanYashchuk committed:
Add refs.nn.functional.layer_norm

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into native-layer-norm

IvanYashchuk committed:
mypy fixes

IvanYashchuk committed:
Remove supports_expanded_weight

IvanYashchuk committed:
xfail gradgrad test and link the issue

IvanYashchuk committed:
Skip jit test because gradgrad is failing

IvanYashchuk committed:
Skip test_correctness_with_reusing_ir

IvanYashchuk commented:
All CI is passing now. Since this PR adds a new OpInfo it revealed a few problems and tests had to be skipped. I submitted an issue https://github.com/pytorch/pytorch/issues/79705.

Except for `torch._refs.var_mean` vs `torch.var_mean` (https://github.com/pytorch/pytorch/pull/79413/files#r899336672) I addressed all the feedback.

@Chillee, @mruberry could you please take another look?

mruberry reviewed (approved):
Great work, @IvanYashchuk! Nice attention to the bug fix and the issue. 

Let's ship this as soon as the tests are happy.

Chillee reviewed (approved):
LGTM, besides the minor nit about `torch.var` and `torch.mean`.

I don't want to block this PR on that though - this PR has been through a bunch of rounds already :)

If it's an issue for using this decomposition in the future we can just modify it to `torch.var`/`torch.mean` then.

IvanYashchuk commented:
@pytorchbot merge -g
